src/arith/iter_affine_map.cc

```diff
@@ -381,31 +534,84 @@ class IterMapRewriter : public ExprMutator {
-    // update the iterator to use the canonicalized form
-    expr.CopyOnWrite()->args = Array<IterSplitExpr>(iters.rbegin(), iters.rend());
-    auto it = sum_fuse_map_.find(expr);
-    if (it != sum_fuse_map_.end()) return it->second;
-    auto mark = IterMark(expr, div(expected_scale, base_scale.value()));
-    IterSplitExpr split(mark, base_scale.value());
-    sum_fuse_map_[expr] = split;
-    return split;
+      if (constraint_to_match) {
+        // match the predicate and mark the iterators in the constraint_to_match as visited
+        // Example: expr = i*9 + j*2 + k, i in [0, 4) j in [0, 5) k in [0, 2)
+        //          predicate = j*2 + k < 9
+        //          then j*2 + k matches the lower two splits of expr
+        for (auto it = constraint_to_match.value()->args.rbegin();
+             it != constraint_to_match.value()->args.rend(); ++it) {
+          size_t k = 0;
+          for (; k < expr->args.size(); ++k) {
+            if (!visited[k] && IterSplitEqual(expr->args[k], *it, false)) {
+              if (analyzer_->CanProveEqual((*it)->scale * expected_scale, expr->args[k]->scale))
+                break;
+            }
+          }
+          if (k == expr->args.size()) return NullOpt;
+          visited[k] = true;
+          flattened_iters.push_back(expr->args[k]);
+        }
+        auto iter = sum_fuse_map_.find(constraint_to_match.value());
+        ICHECK(iter != sum_fuse_map_.end());
+        IterMark iter_matched = iter->second;
+        grouped_iters.emplace_back(iter_matched, expected_scale);
+        expected_scale *= iter_matched->extent;
+        // move forward
+        i += constraint_to_match.value()->args.size();
+      } else {
+        // constraint_to_match not found, skip this iterator
+        visited[j] = true;
+        flattened_iters.push_back(expr->args[j]);
+        grouped_iters.push_back(expr->args[j]);
+        expected_scale *= expr->args[j]->extent;
+        ++i;
+      }
+    }
+    // Get the flattened form and structured form
+    // both forms have splits from outermost to innermost
+    IterSumExpr structured_form = expr, flattened_form = expr;
+    flattened_form.CopyOnWrite()->args =
+        Array<IterSplitExpr>(flattened_iters.rbegin(), flattened_iters.rend());
+    structured_form.CopyOnWrite()->args =
+        Array<IterSplitExpr>(grouped_iters.rbegin(), grouped_iters.rend());
+    auto it = sum_fuse_map_.find(flattened_form);
+    if (it != sum_fuse_map_.end()) {
+      // old iter
+      return IterSplitExpr(it->second, base_scale.value());
+    } else {
+      // new iter, form a new mark
+      IterMark mark = IterMark(structured_form, div(expected_scale, base_scale.value()));
+      sum_fuse_map_[flattened_form] = mark;
+      flattened_map_[structured_form] = flattened_form;
+      return IterSplitExpr(mark, base_scale.value());
+    }
   }
 
   bool CanProveDivisible(const PrimExpr& lhs, const PrimExpr& rhs) {
+    if (analyzer_->CanProveEqual(lhs, rhs)) return true;
```

Comment:
this is not needed. Mainly because the const integer comparison is a faster path while CanProveEqual contains a slower path